### PR TITLE
Preprocessing

### PR DESCRIPTION
**Pull request by erenaydoslu**
Created almost all of the preprocessing code. Should we put `data_downloader.py` here as well?

There are some breaking changes in `tokenize_fetures` and `encode_labels` methods. Both of them also require the save paths now.

### CODE REVIEW

**Comment by daniel-chou-rainho** [comment]
Let me know if my assumptions are wrong.

# Asuming the package exports one function to be called inside model-training or model-service, which does all the pre-processing end-to-end

There a primarily 3 files that encapsulate the preprocessing logic:
`data_split.py`, `encode_labels.py`, and `tokenize_features.py`.

Why do you seperate the functionality into 3 different files?
Because this is forcing you to save intermediary files at the end of each file, adding a lot of boilerplate, and it's not like we can inspect the intermediary files when using the lib-ml as a package, presumable, nor would we want to, presumably.

From the original notebook:

```py
# Load Dataset
train = [line.strip() for line in open("/kaggle/input/dl-dataset/DL Dataset/train.txt", "r").readlines()[1:]]
raw_x_train = [line.split("\t")[1] for line in train]
raw_y_train = [line.split("\t")[0] for line in train]

test = [line.strip() for line in open("/kaggle/input/dl-dataset/DL Dataset/test.txt", "r").readlines()]
raw_x_test = [line.split("\t")[1] for line in test]
raw_y_test = [line.split("\t")[0] for line in test]

val=[line.strip() for line in open("/kaggle/input/dl-dataset/DL Dataset/val.txt", "r").readlines()]
raw_x_val=[line.split("\t")[1] for line in val]
raw_y_val=[line.split("\t")[0] for line in val]

# Tokenize Dataset
tokenizer = Tokenizer(lower=True, char_level=True, oov_token='-n-')
tokenizer.fit_on_texts(raw_x_train + raw_x_val + raw_x_test)
char_index = tokenizer.word_index
sequence_length=200
x_train = pad_sequences(tokenizer.texts_to_sequences(raw_x_train), maxlen=sequence_length)
x_val = pad_sequences(tokenizer.texts_to_sequences(raw_x_val), maxlen=sequence_length)
x_test = pad_sequences(tokenizer.texts_to_sequences(raw_x_test), maxlen=sequence_length)

encoder = LabelEncoder()

y_train = encoder.fit_transform(raw_y_train)
y_val = encoder.transform(raw_y_val)
y_test = encoder.transform(raw_y_test)
```

From the original notebook, the only thing you would have to change is to make this entire thing one function, with input parameter(s) the dataset (just one dataset, no split of 3 I guess, since we are just making inferences in the end), and then output just 2 things: the tokenized features (x) and the encoded labels (y).

# Assuming the package exports multiple functions, one for each file

Assuming one file one function, then right now from what I understand, when you call a function it is just saved as a file, but not returned?

**Comment by daniel-chou-rainho** [comment]
In the first case the code would become smt like this:

```py
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from sklearn.preprocessing import LabelEncoder

def process_data(data):
    # Splitting the data into features and labels
    raw_x = [line.split("\t")[1] for line in data]
    raw_y = [line.split("\t")[0] for line in data]

    # Tokenizing the text data
    tokenizer = Tokenizer(lower=True, char_level=True, oov_token='-n-')
    tokenizer.fit_on_texts(raw_x)
    sequence_length = 200
    x = pad_sequences(tokenizer.texts_to_sequences(raw_x), maxlen=sequence_length)

    # Encoding the labels
    encoder = LabelEncoder()
    y = encoder.fit_transform(raw_y)

    return x, y
```

**Comment by erenaydoslu** [comment]
I assumed we would still be using `dvc` in `model-training` that's why I save the files instead of returning them.

> In the first case the code would become smt like this:
> 
> ```python
> from keras.preprocessing.text import Tokenizer
> from keras.preprocessing.sequence import pad_sequences
> from sklearn.preprocessing import LabelEncoder
> 
> def process_data(data):
>     # Splitting the data into features and labels
>     raw_x = [line.split("\t")[1] for line in data]
>     raw_y = [line.split("\t")[0] for line in data]
> 
>     # Tokenizing the text data
>     tokenizer = Tokenizer(lower=True, char_level=True, oov_token='-n-')
>     tokenizer.fit_on_texts(raw_x)
>     sequence_length = 200
>     x = pad_sequences(tokenizer.texts_to_sequences(raw_x), maxlen=sequence_length)
> 
>     # Encoding the labels
>     encoder = LabelEncoder()
>     y = encoder.fit_transform(raw_y)
> 
>     return x, y
> ```

This could work, the only restriction would be that we would have to do the train-val-test split after preprocessing the data. We don't want to fit the Tokenizer separately on train, val, and test. 

Plus, we would still have to have another function to be used in `model-service` as we won't have labels there.

I could have two separate files for `model-training` and `model-service` that do all the preprocessing in one method and return the results without saving them. Let me know what you think.

**Comment by daniel-chou-rainho** [comment]
> I assumed we would still be using dvc in model-training that's why I save the files instead of returning them.

Makes sense, I didn't think about using DVC in model-training, because I thought the pre-processing would just become 1 function/stage. But I guess it makes sense to keep these functions/stages seperate to inspect intermediary files when running the pipeline inside model-training.

Regarding what these functions return, since they will presumably be used by both model-training and model-service:
In model-training, where we use dvc, we will care about intermediary files, but I don't think we will care about intermediary files inside model-service.

To that extent, the functions could just return the parameters, and then in the case of model-service, it will just continue with calling other functions on it, but for model/training there will be some extra code after the execution of a lib-ml function to save the output as a file, which is read from for the input for the next stage.

**Comment by samuelribaric** [comment]
> Created almost all of the preprocessing code. Should we put `data_downloader.py` here as well?
> 
> There are some breaking changes in `tokenize_fetures` and `encode_labels` methods. Both of them also require the save paths now.

I feel like it makes the most sense to put `data_downloader` there, maybe in the pipeline a simple check would suffice that checks a local folder for the data and if not found it initiates data_downloader?

**Review by samuelribaric — APPROVED** [review]
lgtm, might need more modifications in the future, but for now it looks like a logical start that we can modify accordingly when we get a better grasp of how model-training and model-service will depend on lib-ml.
